simulators/optimism/devnet/devnet.go

```diff
@@ -0,0 +1,205 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/big"
+	"os"
+
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/hive/hivesim"
+)
+
+var (
+	// parameters used for signing transactions
+	chainID  = big.NewInt(901)
+	gasPrice = big.NewInt(30 * params.GWei)
+
+	// would be nice to use a networkID that's different from chainID,
+	// but some clients don't support the distinction properly.
+	networkID = big.NewInt(901)
+)
+
+func bytesSource(data []byte) func() (io.ReadCloser, error) {
+	return func() (io.ReadCloser, error) {
+		return ioutil.NopCloser(bytes.NewReader(data)), nil
+	}
+}
+
+type Devnet struct {
+	t *hivesim.T
+
+	eth1 *Eth1Node
+	l2   *L2Node
+	op   *OpNode
+	l2os *L2OSNode
+
+	genesisTimestamp   string
+	withdrawerBytecode string
+	l1BlockBytecode    string
+
+	l2OutputOracle string
+	optimismPortal string
+
+	nodes map[string]*hivesim.ClientDefinition
+	ctx   context.Context
+}
+
+func (d *Devnet) Start() {
+	clientTypes, err := d.t.Sim.ClientTypes()
+	if err != nil {
+		d.t.Fatal(err)
+	}
+	var eth1, l2, op, l2os *hivesim.ClientDefinition
+	for _, client := range clientTypes {
+		fmt.Println(client.HasRole("ops-l1"))
+		if client.HasRole("ops-l1") {
+			eth1 = client
+		}
+		if client.HasRole("ops-l2") {
+			l2 = client
+		}
+		if client.HasRole("ops-opnode") {
+			op = client
+		}
+		if client.HasRole("ops-l2os") {
+			l2os = client
+		}
+	}
+
+	if eth1 == nil || l2 == nil || op == nil || l2os == nil {
+		d.t.Fatal("ops-l1, ops-l2, ops-opnode, ops-l2os required")
+	}
+
+	// Generate genesis for execution clients
+	//    eth1Genesis := setup.BuildEth1Genesis(config.TerminalTotalDifficulty, uint64(eth1GenesisTime))
+	//    eth1ConfigOpt := eth1Genesis.ToParams(depositAddress)
+	//    eth1Bundle, err := setup.Eth1Bundle(eth1Genesis.Genesis)
+	//    if err != nil {
+	//            t.Fatal(err)
+	//    }
+	var eth1ConfigOpt, eth1Bundle hivesim.Params
+	execNodeOpts := hivesim.Params{
+		"HIVE_CATALYST_ENABLED": "1",
+		"HIVE_LOGLEVEL":         os.Getenv("HIVE_LOGLEVEL"),
+		"HIVE_NODETYPE":         "full",
+	}
+	executionOpts := hivesim.Bundle(eth1ConfigOpt, eth1Bundle, execNodeOpts)
+
+	opts := []hivesim.StartOption{executionOpts}
+	d.eth1 = &Eth1Node{d.t.StartClient(eth1.Name, opts...)}
+
+	d.nodes["ops-l1"] = eth1
+	d.nodes["ops-l2"] = l2
+	d.nodes["ops-opnode"] = op
+	d.nodes["ops-l2os"] = l2os
+}
+
+func (d *Devnet) Wait() error {
+	// TODO: wait until rpc connects
+	client := ethclient.NewClient(d.eth1.Client.RPC())
+	_, err := client.ChainID(d.ctx)
+	return err
+}
+
+func (d *Devnet) DeployL1() error {
+	execInfo, err := d.eth1.Client.Exec("deploy.sh")
+	fmt.Println(execInfo.Stdout)
+	return err
+}
+
+func (d *Devnet) Cat(path string) (string, error) {
+	execInfo, err := d.eth1.Client.Exec("cat.sh", path)
```

Comment:
This can just be `cat`, assuming that `Client.Exec` calls `docker exec` under the hood.

Reply:
Unfortunately the hive `Client` wrapper further restricts which executables can be called from the Simulation API.

https://github.com/ethereum/hive/blob/21c5c3249fd2ecf7ae94dbe93bab9e8e1c4ec932/docs/clients.md#hive-bin

It doesn't expose the underlying `docker.Client` to bypass the restriction either.